zenith_utils/src/pq_proto.rs

```diff
@@ -169,44 +164,53 @@ impl FeStartupMessage {
             bail!("invalid message length");
         }
 
-        let version = stream.read_u32::<BE>()?;
-        let kind = match version {
-            CANCEL_REQUEST_CODE => StartupRequestCode::Cancel,
-            NEGOTIATE_SSL_CODE => StartupRequestCode::NegotiateSsl,
-            NEGOTIATE_GSS_CODE => StartupRequestCode::NegotiateGss,
-            _ => StartupRequestCode::Normal,
-        };
+        let request_code = stream.read_u32::<BE>()?;
 
         // the rest of startup packet are params
         let params_len = len - 8;
         let mut params_bytes = vec![0u8; params_len];
         stream.read_exact(params_bytes.as_mut())?;
 
-        // Then null-terminated (String) pairs of param name / param value go.
-        let params_str = str::from_utf8(&params_bytes).unwrap();
-        let params = params_str.split('\0');
-        let mut params_hash: HashMap<String, String> = HashMap::new();
-        for pair in params.collect::<Vec<_>>().chunks_exact(2) {
-            let name = pair[0];
-            let value = pair[1];
-            if name == "options" {
-                // deprecated way of passing params as cmd line args
-                for cmdopt in value.split(' ') {
-                    let nameval: Vec<&str> = cmdopt.split('=').collect();
-                    if nameval.len() == 2 {
-                        params_hash.insert(nameval[0].to_string(), nameval[1].to_string());
+        // Parse params depending on request code
+        let message = match request_code {
+            CANCEL_REQUEST_CODE => {
+                let parse_i32_big_endian = |byte_slice: &[u8]| {
+                    let byte_array: [u8; 4] = byte_slice.try_into().unwrap();
+                    i32::from_be_bytes(byte_array)
+                };
+
+                ensure!(params_len == 8, "expected 8 bytes for CancelRequest params");
+                let backend_pid = parse_i32_big_endian(&params_bytes[0..4]);
+                let cancel_key = parse_i32_big_endian(&params_bytes[4..8]);
+                FeInitialMessage::CancelRequest(backend_pid, cancel_key)
+            }
+            NEGOTIATE_SSL_CODE => FeInitialMessage::SSLRequest,
+            NEGOTIATE_GSS_CODE => FeInitialMessage::GSSENCRequest,
+            _ => {
+                // Then null-terminated (String) pairs of param name / param value go.
+                let params_str = str::from_utf8(&params_bytes).unwrap();
+                let params = params_str.split('\0');
+                let mut params_hash: HashMap<String, String> = HashMap::new();
+                for pair in params.collect::<Vec<_>>().chunks_exact(2) {
+                    let name = pair[0];
+                    let value = pair[1];
```

Comment:
I think you can simplify this with something like
```
while let Some(name) = params.next() {
    let value = params.next().ok_or_else(|| anyhow!("bad params"))?;
    //insert into params_hash
}

Reply:
It makes the code shorter but turns the iterator into a mutable var, which adds a different kind of complexity. How about `for (name, value) in params.tuples()`? Is `Itertools` [stable enough](https://github.com/rust-itertools/itertools/blob/master/CHANGELOG.md) to depend on?

Reply:
Seems like a cool crate, but is there a way to use it so we can detect `num_params % 2 == 1` with it? Also adding new dependencies is :(

Reply:
Nope :(

So, I wasn't planning on changing this code (my change is just indentation) but now I notice that `chunks_exact` just means "if there's leftover, we'll keep that in the iterator so you can check later yourself", and we're not checking that now. I think your suggestion is the simplest way to actually check that :)